docs/remote-attestation.md

```diff
@@ -0,0 +1,148 @@
+     pair
+   - **TEE Provider** creates a signature of the **TEE Platform**'s firmware key
+1. **Client** generates an
+   [X25519](https://datatracker.ietf.org/doc/html/rfc7748) _Ephemeral_ key pair
+1. **Client** sends a `ClientHello` message to the **Server**
+   - Which includes a random string
+1. **Server** generates an
+   [X25519](https://datatracker.ietf.org/doc/html/rfc7748) _Ephemeral_ key pair
+1. **Server** sends a request for the `AttestationReport` to the **TEE
+   Platform**
+   - This request contains
+     [SHA-256](https://datatracker.ietf.org/doc/html/rfc6234) hash of the
+     **Server**’s _Signing_ public key
+1. **TEE Platform** generates an `AttestationReport`
+   - **TEE Platform** signs the report using its firmware key that was signed by
+     the **TEE Provider** and could be verified by the **Client**
+   - The signed `AttestationReport` also gives the **Client** information about
+     the TEE itself (its version and etc.) and about the code that is running
+     inside the **TEE Platform**
+   - Since the hash of the **Server**’s _Signing_ public key was provided in the
+     request, it is also signed by the **TEE Platform**
+1. **Server** sends `ServerIdentity` to the **Client** which contains:
+   - **Server**’s _Ephemeral_ public key
+   - New random string
+   - _Transcript_: [SHA-256](https://datatracker.ietf.org/doc/html/rfc6234) hash
+     of the concatenated `ClientHello` and current `ServerIdentity` (excluding
+     the _Transcript_) signed by the **Server**'s _Signing_ public key
+   - **Server**’s _Signing_ public key
+   - `AttestationReport` signed by the **TEE Platform**'s hardware key, which
+     includes a hash of the **Server**’s _Signing_ public key
+   - Corresponding **TEE Provider**’s certificate that is signed by the **TEE
+     Provider**’s _Root_ key
+1. **Client** validates `ServerIdentity`
+   - If the corresponding `AttestationReport` is not valid, then the **Client**
+     closes the connection and aborts the protocol
+1. **Client** sends `ClientIdentity` to the **Server** which contains:
+   - **Client**’s _Ephemeral_ public key
+   - New random string
+   - _Transcript_: [SHA-256](https://datatracker.ietf.org/doc/html/rfc6234) hash
+     of the concatenated `ClientHello`, `ServerIdentity` and current
+     `ClientIdentity` (excluding the _Transcript_) signed by the **Client**'s
+     _Signing_ public key
+   - **Client**’s _Signing_ public key
+1. **Client** and **Server** use both _Ephemeral_ public keys to create an
+   [X25519](https://datatracker.ietf.org/doc/html/rfc7748) _Shared Secret_
+1. **Client** and **Server** derive _Session_ keys from the _Shared Secret_ and
+   both _Ephemeral_ public keys using a _Key Derivation Function_
+   ([HKDF](https://datatracker.ietf.org/doc/html/rfc5869))
+   - Each side generates 2
+     [AES-256-GCM](https://datatracker.ietf.org/doc/html/rfc5288) _Session_
+     keys:
+     - Server Session Key
+     - Client Session Key
```

Comment:
Each side has both of them? I'm a bit confused by this, isn't AES 256 symmetric? Do we use two symmetric keys?

Reply:
These are 2 different symmetric keys (see a previous response).